pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package infracluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	mapiv1beta1 "github.com/openshift/api/machine/v1beta1"
+	"gopkg.in/yaml.v2"
```

Comment:
This may not be the cause of the problems, but use `sig.k8s.io/yaml` instead here. The marshalling behavior is different between the two libraries.

https://github.com/openshift/cluster-capi-operator/blob/main/pkg/controllers/infracluster/vsphere.go#L34